lib/bolt/pal/yaml_plan/step/resources.rb

```diff
@@ -0,0 +1,105 @@
+# frozen_string_literal: true
+
+module Bolt
+  class PAL
+    class YamlPlan
+      class Step
+        class Resources < Step
+          def self.allowed_keys
+            super + Set['resources']
+          end
+
+          def self.required_keys
+            Set['target']
+          end
+
+          def initialize(step_body)
+            super
+            @resources = step_body['resources']
+            @normalized_resources = normalize_resources(@resources)
+          end
+
+          def self.validate(body, step_number)
+            super
+
+            body['resources'].each do |resource|
+              if resource['type'] || resource['title']
+                if !resource['type']
+                  err = "Resource declaration must include type key if title key is set"
+                  raise step_error(err, body['name'], step_number)
+                elsif !resource['title']
+                  err = "Resource declaration must include title key if type key is set"
+                  raise step_error(err, body['name'], step_number)
+                elsif resource['type'].is_a?(CodeLiteral)
```

Comment:
Not quite sure how to detect something like `$foo` when `type: $foo` as it is a `BareString`
```
cas@cas-ThinkPad-T460p:~/working_dir/bolt$ bolt plan run test::eval

From: /home/cas/working_dir/bolt/lib/bolt/pal/yaml_plan/step/resources.rb @ line 26 Bolt::PAL::YamlPlan::Step::Resources.validate:

    22: def self.validate(body, step_number)
    23:   super
    24: 
    25:   body['resources'].each do |resource|
 => 26:     require 'pry'; binding.pry
    27:     if resource['type'] || resource['title']
    28:       if !resource['type']
    29:         err = "Resource declaration must include type key if title key is set"
    30:         raise step_error(err, body['name'], step_number)
    31:       elsif !resource['title']
    32:         err = "Resource declaration must include title key if type key is set"
    33:         raise step_error(err, body['name'], step_number)
    34:       elsif resource['type'].is_a?(CodeLiteral)
    35:         err = "Code expression '#{resource['type'].value}' cannot be used as a resource type"
    36:         raise step_error(err, body['name'], step_number)
    37:       end
[1] pry(Bolt::PAL::YamlPlan::Step::Resources)> resource['type']
=> #<Bolt::PAL::YamlPlan::BareString:0x0000559792fb4620 @value="$foo">
```
Example yaml plan
```
parameters:
  foo:
    type: Optional[String]
    description: foo
    default: exec
steps:
  - name: apply_resource
    target: localhost
    resources:
      - type: $foo
        title: echo hi > foo
        parameters:
          path: '/usr/bin:/usr/sbin:/bin'
          provider: shell
          cwd: '/home/cas/working_dir/bolt'
      - type: notify
        title: 'foo'
  - name: cat
    eval: |
      file::read('/home/cas/working_dir/bolt/foo')

return: [$apply_resource, $cat]
```